src/m365/entra/commands/pim/pim-role-assignment-add.ts

```diff
@@ -0,0 +1,319 @@
+      }
+    });
+    this.optionSets.push({
+      options: ['administrativeUnitId', 'applicationId'],
+      runsWhen: (args) => {
+        return args.options.administrativeUnitId !== undefined || args.options.applicationId !== undefined;
+      }
+    });
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    let action = 'adminAssign';
+
+    try {
+      const token = auth.connection.accessTokens[auth.defaultResource].accessToken;
+      const isAppOnlyAccessToken = accessToken.isAppOnlyAccessToken(token);
+      if (isAppOnlyAccessToken) {
+        throw 'When running with application permissions either userId, userName, groupId or groupName is required';
+      }
+
+      const roleDefinitionId = await this.getRoleDefinitionId(args.options, logger);
+      const principalId = await this.getPrincipalId(args.options, logger);
+
+      if (!args.options.userId && !args.options.userName && !args.options.groupId && !args.options.groupName) {
+        action = 'selfActivate';
+      }
+
+      const requestOptions: CliRequestOptions = {
+        url: `${this.resource}/v1.0/roleManagement/directory/roleAssignmentScheduleRequests`,
+        headers: {
+          'accept': 'application/json;odata.metadata=none'
+        },
+        responseType: 'json',
+        data: {
+          principalId: principalId,
+          roleDefinitionId: roleDefinitionId,
+          directoryScopeId: this.getDirectoryScope(args.options),
+          action: action,
+          justification: args.options.justification,
+          scheduleInfo: {
+            startDateTime: args.options.startDateTime,
+            expiration: {
+              duration: this.getDuration(args.options),
+              endDateTime: args.options.endDateTime,
+              type: this.getExpirationType(args.options)
+            }
+          },
+          ticketInfo: {
+            ticketNumber: args.options.ticketNumber,
+            ticketSystem: args.options.ticketSystem
+          }
+        }
+      };
+
+      const response = await request.post<UnifiedRoleAssignmentScheduleRequest>(requestOptions);
+
+      await logger.log(response);
+    }
+    catch (err: any) {
+      this.handleRejectedODataJsonPromise(err);
+    }
+  }
+
+  private async getRoleDefinitionId(options: Options, logger: Logger): Promise<string> {
+    if (options.roleDefinitionId) {
+      return options.roleDefinitionId;
+    }
+
+    if (this.verbose) {
+      await logger.logToStderr(`Retrieving role definition by its name '${options.roleDefinitionName}'`);
+    }
+
+    const role = await roleDefinition.getRoleDefinitionByDisplayName(options.roleDefinitionName!);
+    return role.id!;
+  }
+
+  private async getPrincipalId(options: Options, logger: Logger): Promise<string> {
+    let principalId = options.userId;
```

Comment:
Let's just immediately return userId if it's filled, same for groupId.

No need to create a variable and reassign it all the time. Just return what you retrieve as well.